### PR TITLE
feat: create testimonial

### PR DESCRIPTION
**Pull request by CFCIfe**
## What does the PR do?

This PR is about creating testimonials for logged in users. The endpoint returns errors if:

- [x] User is not logged in.
- [x] User does not exist.
- [x] Validation errors on the user input is done

How should this be manually tested?

1. Login to a user account, and get the access token

2. Make a POST request to the `\api\testimonials` endpoint with the following payload:

```json
Authorization: Bearer <access_token>

{
  "name": "John Doe",
  "content": "The service is great. I love it. I hope to come back soon."
}
```

Successful Response

On a successful testimonial creation, the API should return a 201 CREATED status code.

```json
{
      "status": "success",
      "message": "Testimonial created successfully",
      "data": {
            "user_id": "a9f4b3b0-4b3b-4b3b-4b3b-4b3b4b3b4b3b",
            "name": "John Doe",
            "content": "The service is great. I love it. I hope to come back soon."
            "created_at":"2021-08-01T12:00:00Z"
      }
}
```

image

Authentication Error:

When the current user is not properly authenticated to change the organization detail:

```json
{
  "status": "error",
  "message": "User is currently unauthorized, kindly authenticate to continue",
  "status_code": 401
}
```
no token image

Invalid UserID Error

On Invalid error, if the user is not found with the provided id:

```json
{
  "status": "Error",
  "message": "user not found",
  "status_code": 404
}
```
wrong user ID image

Validation:

If any required field is missing or invalid, the response should return a 400 Bad Request status code with appropriate validation error messages.

```json
{
  "status": "error",
  "message": ["name must be a string", "content must be a string"],
  "status_code": "400"
}
```
validation_error_image

Task

- [x] Create a POST endpoint PATCH /api/v1/testimonials/ to create testimonials for logged in users.
- [x] Implement server-side logic to handle successful and failed update request.
- [x] Ensure the user has the proper authentication to create testimonials.
- [x] Update the testimonials details in the database.
- [x] Write tests for creating testimonials.
- [x] Update API documentation to reflect the new endpoint and its usage.

Reference Issue: [Link to Issue](https://github.com/hngprojects/hng_boilerplate_nestjs/issues/206)

### CODE REVIEW

**Comment by CFCIfe** [comment]
Converting to draft pending when the Auth Gauard goes to dev branch

**Review by davidolanrewaju — APPROVED** [review]

**Review by davidolanrewaju — COMMENTED** [review]
Please let your PR sample follow PR #351 pattern

File: db/migrations/1721779124878-migration.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class Migration1721779124878 implements MigrationInterface {
+  name = 'Migration1721779124878';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(
+      `CREATE TABLE "abstract_base_entity" ("id" uuid NOT NULL DEFAULT uuid_generate_v4(), "created_at" TIMESTAMP NOT NULL DEFAULT now(), "updated_at" TIMESTAMP NOT NULL DEFAULT now(), CONSTRAINT "PK_723c213727b1038d476972ce2f2" PRIMARY KEY ("id"))`
```
Comment (Homoakin619) [inline db/migrations/1721779124878-migration.ts:8]: Remove this migration, abstract base entity is just an abstract entity and does not need a table in the database, It was created to be extended by other entities.

File: db/migrations/1721779124878-migration.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    await queryRunner.query(
+      `ALTER TABLE "testimonial" ADD CONSTRAINT "FK_8455ac0da5f4cc082831516abb1" FOREIGN KEY ("userId") REFERENCES "user"("id") ON DELETE NO ACTION ON UPDATE NO ACTION`
+    );
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`ALTER TABLE "testimonial" DROP CONSTRAINT "FK_8455ac0da5f4cc082831516abb1"`);
+    await queryRunner.query(`DROP TABLE "testimonial"`);
+    await queryRunner.query(`DROP TABLE "user"`);
+    await queryRunner.query(`DROP TABLE "abstract_base_entity"`);
```
Comment (Homoakin619) [inline db/migrations/1721779124878-migration.ts:25]: remove the drop query here also

File: src/modules/testimonials/dto/create-testimonial.dto.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  @IsNotEmpty()
+  @ApiProperty({
+    type: String,
+    description: 'Content of the testimonial',
+    example: 'I am very happy with the service provided by the company',
+  })
+  content: string;
+}
+
+export class CreateTestimonialResponseDto {
```
Comment (Homoakin619) [inline src/modules/testimonials/dto/create-testimonial.dto.ts:24]: export this to it's separate file

File: src/modules/testimonials/dto/create-testimonial.dto.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  'status': string;
+  'message': string;
+  'data': TestimonialData;
+}
+
+interface TestimonialData {
```
Comment (Homoakin619) [inline src/modules/testimonials/dto/create-testimonial.dto.ts:30]: export this to a new file as well, probably in an interface folder, and rename to TestimonialInterface or TestimonialDataInterface

File: src/modules/testimonials/testimonials.service.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+@Injectable()
+export class TestimonialsService {
+  constructor(
+    @InjectRepository(Testimonial)
+    private readonly testimonialRepository: Repository<Testimonial>,
+
+    @InjectRepository(User)
+    private readonly userRepository: Repository<User>
+  ) {}
+  async create(createTestimonialDto: CreateTestimonialDto, userId: string) {
```
Comment (Homoakin619) [inline src/modules/testimonials/testimonials.service.ts:17]: could you make this name more descriptive like having createTestimonial

File: src/modules/testimonials/testimonials.service.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+import { Testimonial } from './entities/testimonials.entity';
+
+@Injectable()
+export class TestimonialsService {
+  constructor(
+    @InjectRepository(Testimonial)
+    private readonly testimonialRepository: Repository<Testimonial>,
+
+    @InjectRepository(User)
+    private readonly userRepository: Repository<User>
```
Comment (Homoakin619) [inline src/modules/testimonials/testimonials.service.ts:15]: do away with this

File: src/modules/testimonials/testimonials.service.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  ) {}
+  async create(createTestimonialDto: CreateTestimonialDto, userId: string) {
+    const { content, name } = createTestimonialDto;
+
+    const user = await this.userRepository.findOneBy({ id: userId });
```
Comment (Homoakin619) [inline src/modules/testimonials/testimonials.service.ts:20]: there is a method in the user service that takes care of this, kindly reference the method and don't re-invent the wheel. 

UserService.getUserRecord

File: db/migrations/1721779124878-migration.ts (hunk 1 above)
Comment (CFCIfe) [reply db/migrations/1721779124878-migration.ts:8]: Removed query

File: db/migrations/1721779124878-migration.ts (hunk 2 above)
Comment (CFCIfe) [reply db/migrations/1721779124878-migration.ts:25]: Removed drop query

File: src/modules/testimonials/dto/create-testimonial.dto.ts (hunk 4 above)
Comment (CFCIfe) [reply src/modules/testimonials/dto/create-testimonial.dto.ts:30]: Moved

File: src/modules/testimonials/dto/create-testimonial.dto.ts (hunk 3 above)
Comment (CFCIfe) [reply src/modules/testimonials/dto/create-testimonial.dto.ts:24]: done

File: src/modules/testimonials/testimonials.controller.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+export class TestimonialsController {
+  constructor(private readonly testimonialsService: TestimonialsService) {}
+
+  @Post()
+  @ApiOperation({ summary: 'Create Testimonials' })
+  @ApiResponse({ status: 201, description: 'Testimonial created successfully' })
+  @ApiResponse({ status: 404, description: 'User not found' })
+  @ApiResponse({ status: 401, description: 'Unauthorized' })
+  @ApiResponse({ status: 500, description: 'Internal Server Error' })
+  @UseGuards(AuthGuard)
```
[the remark this answers is not in the document]
Comment (CFCIfe) [reply src/modules/testimonials/testimonials.controller.ts:21]: Removed auth guard

File: src/modules/testimonials/testimonials.service.ts (hunk 5 above)
Comment (CFCIfe) [reply src/modules/testimonials/testimonials.service.ts:17]: Renamed to createTestimonial

File: src/modules/testimonials/testimonials.service.ts (hunk 7 above)
Comment (CFCIfe) [reply src/modules/testimonials/testimonials.service.ts:20]: Done

File: src/modules/testimonials/testimonials.service.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+import { CreateTestimonialDto } from './dto/create-testimonial.dto';
+import { Testimonial } from './entities/testimonials.entity';
+
+@Injectable()
+export class TestimonialsService {
+  constructor(
+    @InjectRepository(Testimonial)
+    private readonly testimonialRepository: Repository<Testimonial>
+  ) {}
+  async createTestimonial(createTestimonialDto: CreateTestimonialDto, user) {
```
Comment (Homoakin619) [inline src/modules/testimonials/testimonials.service.ts:13]: add a try catch block here to handle exceptions that might occur

Comment (CFCIfe) [reply src/modules/testimonials/testimonials.service.ts:13]: When you implement the try…catch in the controller it overrides whatever error your throwing from the service. @alphacodez . I used to have it but removed it because all errors become 500

Comment (CFCIfe) [reply src/modules/testimonials/testimonials.service.ts:13]: Done in the service file

**Review by Homoakin619 — APPROVED** [review]

File: src/main.ts [hunk 10]
```diff
@@ -29,25 +30,23 @@ async function bootstrap() {
...
-    .setTitle('<project-title-here>')
+    .setTitle('Remote Bingo')
     .setDescription('<project-description-here>')
     .setVersion('1.0')
     .addBearerAuth()
     .build();
 
   const document = SwaggerModule.createDocument(app, options);
-  SwaggerModule.setup('api/docs', app, document);
+  SwaggerModule.setup('api', app, document);
```
Comment (buka4rill) [inline src/main.ts:45]: Why did you change this again? Was this at least approved by the team lead? 
cc: @Heba-WebDev

File: src/main.ts [hunk 11]
```diff
@@ -29,25 +30,23 @@ async function bootstrap() {
...
   await app.listen(port);
 
-  logger.log({ message: 'server started 🚀', port, url: `http://localhost:${port}/api/v1` });
+  logger.log({ message: 'server started 🚀', port, url: `http://localhost:${port}/api` });
 }
-bootstrap().catch(err => {
-  console.error('Error during bootstrap', err);
-  process.exit(1);
-});
+bootstrap();
```
Comment (buka4rill) [inline src/main.ts:52]: Why are you reverting this?

File: src/main.ts (hunk 10 above)
Comment (CFCIfe) [reply src/main.ts:45]: Someone overwrote that file and i blindly reverted to an older commit. I'm using the current one now

Comment (buka4rill) [reply src/main.ts:45]: I think it was decided that it would be api/docs

File: src/modules/testimonials/dto/create-testimonial-response.dto.ts [hunk 12]
```diff
@@ -0,0 +1,7 @@
+import { TestimonialData } from '../interfaces/testimonials.interface';
+
+export class CreateTestimonialResponseDto {
```
[the remark this answers is not in the document]
Comment (CFCIfe) [reply src/modules/testimonials/dto/create-testimonial-response.dto.ts:5]: Done

**Comment by CFCIfe** [comment]
@buka4rill I have reviewed all the changes. Thanks

**Review by Cobidex — APPROVED** [review]

File: src/modules/testimonials/dto/create-testimonial-response.dto.ts [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  })
+  'status': string;
+  @IsString()
+  @IsNotEmpty()
+  @ApiProperty({
+    type: String,
+    description: 'Message',
+    example: 'Testimonial created successfully',
+  })
+  'message': string;
```
Comment (Homoakin619) [inline src/modules/testimonials/dto/create-testimonial-response.dto.ts:21]: remove the quote from this 
property

File: src/modules/testimonials/dto/create-testimonial-response.dto.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+    type: TestimonialData,
+    description: 'Testimonial data',
+    example: {
+      user_id: '1',
+      name: 'John Doe',
+      content: 'I am very happy with the service provided by the company',
+      created_at: new Date(),
+    },
+  })
+  'data': TestimonialData;
```
Comment (Homoakin619) [inline src/modules/testimonials/dto/create-testimonial-response.dto.ts:33]: remove the quote from this property

File: src/modules/testimonials/testimonials.controller.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+  @ApiResponse({ status: 401, description: 'Unauthorized' })
+  @ApiResponse({ status: 500, description: 'Internal Server Error' })
+  async create(
+    @Body() createTestimonialDto: CreateTestimonialDto,
+    @Req() req: Request
+  ): Promise<CreateTestimonialResponseDto> {
+    const { sub: userId } = req?.user as { sub: string };
+
+    const user = await this.userService.getUserRecord({ identifier: userId, identifierType: 'id' });
+
```
Comment (Homoakin619) [inline src/modules/testimonials/testimonials.controller.ts:32]: You don't ever trust a record. What if a user is not found after trying to get the record, that scenario is not handled here kindly do that

File: src/main.ts (hunk 10 above)
Comment (CFCIfe) [reply src/main.ts]: Yes

**Review by buka4rill — APPROVED** [review]

**Comment by CFCIfe** [comment]
@buka4rill all checks passed.
